doxygen/contributor_help_pages/common_pitfalls.md

```diff
@@ -0,0 +1,239 @@
+is not an error (SFINAE) semantics. For instance, for the norm function above,
+SFINAE could be used to limit one signature to work with reverse
+mode autodiff types and one to work with anything else:
+
+```cpp
+template <typename T,
+          require_st_var<T>* = nullptr>
+return_type_t<T> norm(const T&);
+template <typename T,
+          require_not_st_var<T>* = nullptr>
+return_type_t<T> norm(const T&);
+```
+
+SFINAE should be thought of as filters on what functions are visible to the
+compiler when it does name lookup for a specific function. `require_st_var`
+should be read "require the scalar type of the argument to be a `var`". The
+meta-program `require_st_var<T>` will evaluate to a valid type if the scalar
+type of `T` is a `var`. If the scalar type of `T` is not a `var`, then
+`require_st_var<T>` does not evaluate to a valid type, and it is not possible
+to do template substitution on this signature, and the compiler treats it as
+if it does not exist. This is how SFINAE (substitution failure is not an error)
+works. Because the substitution does not work, the signature is ignored.
+This is all built similarly to C++'s
+[std::enable_if](https://en.cppreference.com/w/cpp/types/enable_if)
+template metaprogram.
+
+Again, there are many ways to solve a problem in C++. In particular there
+are cases where simple overloads or template specializations can achieve the
+same thing as the SFINAE template meta-programs. Unless there is a specific
+reason though, new functions in Math should use the SFINAE meta-programs to
+handle different implementations. These are tested to work with the broad
+set of C++ types that the relatively compact set of Stan types map to.
+
+The vast majority of the SFINAE template meta-programs in Math are special
+to Stan or Math. Documentation can be found in the
+[Doxygen](https://mc-stan.org/math/) docs. If functionality is not available,
+it new meta-programs should be added.
+
+
+
+### Reference types
+
+Any time a function takes a vector or matrix type, it needs to be able
+to handle an expression. Some expressions can be expensive to evaluate, so
+when they are evaluated, they should be saved. This includes any arithmetic
+operation, a matrix multiply, or a sum, or something else. Some expressions
+are cheap to evaluate though, any of the Eigen views quality here (transpose,
+block access, segment, etc). In this case, evaluating the expression is not
+necessary -- it would only lead to another allocate and copy.
+
+The Math function `to_ref` is a solution for this. For a vector
+or matrix variable `x`, `to_ref(x)` returns an Eigen reference to the
+input variable `x`. If `x` was an expensive expression, it will be evaluated.
+If it was a cheap expression, the reference type won't evaluate. If it
+isn't an Eigen type, the `to_ref` just forwards.
+
+```cpp
+template <typename T,
+          require_eigen_t<T>* = nullptr>
+auto myfunc(const T& x) {
+  const auto& x_ref = to_ref(x);
+}
+```
+
+### Holders
```

Comment:
Tagging @t4c1 and @andrjohns can you both look this over to make sure it covers the problem and solution well?

Reply:
Looks fine to me.